src/BidiStreamingResponse.php

```diff
@@ -0,0 +1,174 @@
+     *
+     * @param mixed $request The request to write
+     * @throws ValidationException
+     */
+    public function write($request)
+    {
+        if ($this->writesClosed) {
+            throw new ValidationException("Cannot call write() after calling closeWrite().");
+        }
+        if ($this->isComplete) {
+            throw new ValidationException("Cannot call write() after streaming call is complete.");
+        }
+        $this->call->write($request);
+    }
+
+    /**
+     * Write all requests in $requests.
+     *
+     * @param mixed[] $requests An Iterable of request objects to write to the server
+     *
+     * @throws ValidationException
+     * @throws ApiException
+     */
+    public function writeAll($requests = [])
+    {
+        foreach ($requests as $request) {
+            $this->write($request);
+        }
+    }
+
+    /**
+     * Inform the server that no more requests will be written. The write() function cannot be
+     * called after closeWrite() is called.
+     */
+    public function closeWrite()
+    {
+        if ($this->isComplete) {
+            throw new ValidationException(
+                "Cannot call closeWrite() after streaming call is complete."
+            );
+        }
+        if (!$this->writesClosed) {
+            $this->call->writesDone();
+            $this->writesClosed = true;
+        }
+    }
+
+    /**
+     * Read the next response from the server. Returns null if the streaming call completed
+     * successfully. Throws an ApiException if the streaming call failed.
+     *
+     * @return mixed
+     * @throws ValidationException
+     * @throws ApiException
+     */
+    public function read()
+    {
+        if ($this->isComplete) {
+            throw new ValidationException("Cannot call read() after streaming call is complete.");
+        }
+        $resourcesField = $this->resourcesField;
+        if (!is_null($resourcesField)) {
+            if (count($this->pendingResources) === 0) {
+                $response = $this->call->read();
+                if (!is_null($response)) {
+                    $this->pendingResources = array_reverse($response->$resourcesField());
+                }
+            }
+            $result = array_pop($this->pendingResources);
+        } else {
+            $result = $this->call->read();
+        }
+        if (is_null($result)) {
+            $status = $this->call->getStatus();
+            $this->isComplete = true;
+            if (!($status->code == Grpc\STATUS_OK)) {
+                throw new ApiException($status->details, $status->code);
+            }
+        }
+        return $result;
+    }
+
+    /**
+     * Call closeWrite(), and read all responses from the server, until the streaming call is
+     * completed. Throws an ApiException if the streaming call failed.
+     *
+     * @return \Generator|mixed[]
+     * @throws ValidationException
+     * @throws ApiException
+     */
+    public function closeAndReadAll()
```

Comment:
Yes they can. The reason to combine them is that readAll() doesn't really make sense if you *haven't* called closeWrite(). For example with speech, if you write audio and then call read() in a loop without calling closeWrite(), the call to read() will block until the streaming call times out and it returns null. So calling read() before closeWrite() has the potential block and timeout if the server is waiting on more calls to write() before sending anything back, while calling readAll() before closeWrite() is *guaranteed* to timeout (or block forever).

Maybe rename to closeWriteAndReadAll()?